toolchain/common/enum_base.def

```diff
@@ -0,0 +1,96 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Provides an enum type with names for values.
+//
+// Uses should look like:
+//   #define CARBON_ENUM_BASE_NAME MyEnumBase
+//   #define CARBON_ENUM_DEF_PATH "my_xmacros.def"
+//   #include "toolchain/common/enum_base.def"
+//
+//   class MyEnum : public MyEnumBase<MyEnum> {
+//    public:
+//     (any custom APIs)
+//    protected:
+//     using MyEnumBase::MyEnumBase;
+//   };
+//
+// my_xmacros.def will provide:
+//   #ifdef CARBON_ENUM_DEF_PATH
+//   #define MY_XMACRO(Name) CARBON_ENUM_ENTRY(Name)
+//   #endif
+//
+// Specific macro values are available as:
+//   MyEnum::Name()
+//
+// They will be usable in a switch statement, e.g. `case MyEnum::Name():`.
+
+#ifndef CARBON_ENUM_DEF_PATH
+#error "Must define CARBON_ENUM_DEF_PATH"
+#endif
+#ifndef CARBON_ENUM_BASE_NAME
+#error "Must define CARBON_ENUM_BASE_NAME"
+#endif
+
+namespace Internal {
+namespace CARBON_ENUM_BASE_NAME {
+
+inline constexpr llvm::StringLiteral Names[] = {
+#define CARBON_ENUM_ENTRY(Name) #Name,
+#include CARBON_ENUM_DEF_PATH
```

Comment:
Have you checked that `bazel`'s dependency scanning mechanism can cope with this properly?

Reply:
Bazel's may, but others won't. Specifically, ones that use the compiler-provided dependency file will work correctly, but none of the other header-use-scanning tools handle this AFAIK. Certainly, the one that Google uses internally (unsure if its upstreamed as part of Bazel or not) doesn't.

Fundamentally, this is a really dark and problematic corner of the C++ preprocessor that I sadly don't think we should use...

Is there an approach that avoids this?

Reply:
I'd been talking with zygoloid yesterday about options, here's the end result. It doesn't refactor TokenKind, but I'm confident about my ability to do that if we can agree on the basic approach.

Note, bazel handles the files fine regardless -- it does not do include scanning the way blaze might.

Reply:
TBH, I find the new version significantly less easy to understand than the status-quo (before this PR entirely).

I feel like it adds a bunch of complexity, and it makes the documentation significantly more complex.

I don't feel like I even understand the motivation or what problem you're trying to solve by changing from the original approach, so its hard for me to suggest an alternative at this point.

Reply:
> I don't feel like I even understand the motivation or what problem you're trying to solve by changing from the original approach, so its hard for me to suggest an alternative at this point.

What do you mean "original approach"? The status quo, or the original state of this PR?

Versus the status quo, my goal is to eliminate the boilerplate.

Versus the original state, I'm stopping use of the "really dark and problematic corner of the C++ preprocessor that I sadly don't think we should use..."

Reply:
FWIW, I'll add that the status quo boilerplate is creating inconsistencies in the implementations, and it's a touch annoying to be doing copy-paste code.

Reply:
If you're not understanding how rejecting the original approach led to this approach:

The enum base needs some way to invoke each macro three times:

- The #include redirection was addressing this by including the .def file in each spot, but that requires #include.
- The current approach addresses it through X-macros which can be invoked without a #include.

There may be some other way to do this, e.g. by bouncing back and forth, maybe requiring providing EnumBase in a series of 4-ish macros which define portions of the class. Then you could get:

```
// Start the class
ENUM_BASE_PART_1()
// First enum loop
#define MY_ENUM(X) ENUM_ITERATION_PART_1(X)
#include "my_enum.def"
// Next part of the class
ENUM_BASE_PART_2()
// Second enum loop
#define MY_ENUM(X) ENUM_ITERATION_PART_2(X)
#include "my_enum.def"
// Third part of the class
ENUM_BASE_PART_3()
// Final enum loop
#define MY_ENUM(X) ENUM_ITERATION_PART_3(X)
#include "my_enum.def"
// Final part of the class
ENUM_BASE_PART_4()
```

And maybe that's where I'll go next, but TBH I don't think the current approach bothers me much (while a bit tedious to switch, the `\` at EOL doesn't seem that bad to me, although it's apparently a non-starter for you?)

Reply:
I was asked (by @Pixep) how many of these we have that I'm looking at sharing code: I think we have right now ParseNodeKind, ParserState, SemanticsNodeKind, SemanticsBuiltinKind, and TokenKind. DiagnosticKind could easily be migrated (I didn't originally in order to avoid the copy-paste). There might be more; e.g., ValueKind in explorer is being switched to a .def and could maybe share, and there are other enums where we might benefit from a unified approach. And presumably we'll keep getting more enums.

So I'm now at five approaches, and how I see trade-offs:

# 1) Status quo

Advantages:

- The most explicit in behavior (no base class, because you just copy-paste)

Disadvantages:

- Copy-paste boilerplate is difficult to maintain, has already led to skew in behavior.
- Requires re-reading the class definitions to understand behavior, notice differences.
- Currently several .cpp files only exist to return names because of the copy-paste consistency
  - Although possible, it'd look odd putting those in the header under this approach (because others would still want the cpp, increasing burden of understanding the pattern).

# 2) #include redirection

https://github.com/carbon-language/carbon-lang/pull/2453/files/ebdc40b904cd32177ae68aaaca6d7c7e63150c53

Advantages:

- Short reuse (3 LOC) -- particularly comparing with approach 4

Disadvantages:

- #include redirection doesn't work well with Blaze's include scanner (Bazel doesn't support this)
- Concerns from chandlerc that other things may have issues too

# 3) X-macros

https://github.com/carbon-language/carbon-lang/pull/2453/files/64000661bfc0728302209256f4b8d6d236650866

Advantages:

- Allows removal of .def files and using traditional C++ headers.
  - Could also keep the macros in their own header.

Disadvantages:

- Requires changing all .def macros.
  - Requires `/* */` comments.
  - Long macros are a nuisance with end-of-line `\`.

# 4) Repeat #defines

https://github.com/carbon-language/carbon-lang/pull/2453/files/2ad078b608c68b458661bade78ae1c3f881b5e47

Advantages:

- Mostly, I think this lacks disadvantages of other approaches.

Disadvantages:

- Longer boilerplate macro block (1_OF_7, etc).
- clang-format doesn't work on the base class due to the macro separation.

# 5) Generic base class with some #define helpers

https://github.com/carbon-language/carbon-lang/pull/2453/files/cdc06cca9547f7f03a4b1b6db66a4ac6fb6fce4b

Advantages:

- Makes less use of macros than other approaches, just sharing already-defined macros.

Disadvantages:

- Longest boilerplate, requires enum to be defined outside class so that the base class can see it.
  - Longer than approach 4 because things like the enum and name() need to be fully written out separately, rather than benefiting from a macro.